File: database/edit_logger.py
"""
Edit Logger Module

Logs all changes to core tables for sync and audit purposes.
Each edit is stored as an immutable log entry with a UUID for global uniqueness.
"""

import uuid
from datetime import datetime
from typing import Optional, Dict, Any, List


class EditLogger:
    """
    Logs edits to the database for sync and audit tracking.

    Usage:
        logger = EditLogger(cursor, connection)
        logger.log_insert('locations', 123, {'name': 'Central Park', 'lat': 40.785091})
        logger.log_update('locations', 123, 'name', 'Central Park', 'Central Park NYC')
        logger.log_delete('locations', 123, {'name': 'Central Park', ...})
    """

    # Tables that should have edits logged
    TRACKED_TABLES = {
        'locations', 'location_alternate_names', 'location_tags',
        'websites', 'website_urls', 'website_locations', 'website_tags',
        'events', 'event_occurrences', 'event_urls', 'event_tags',
        'tags', 'tag_rules'
    }

    def __init__(self, cursor, connection, source: str = 'local', editor_info: str = None):
        """
        Initialize the edit logger.

        Args:
            cursor: Database cursor
            connection: Database connection
            source: Origin of edits ('local', 'website', or 'crawl')
            editor_info: Additional context (e.g., 'crawl_run:123', 'admin')
        """
        self.cursor = cursor
        self.connection = connection
        self.source = source
        self.editor_info = editor_info
        self.user_id = None
        self.editor_ip = None
        self.editor_user_agent = None

    def set_user_context(self, user_id: int = None, ip: str = None, user_agent: str = None):
        """Set user context for attribution."""
        self.user_id = user_id
        self.editor_ip = ip
        self.editor_user_agent = user_agent

    def _generate_uuid(self) -> str:
        """Generate a UUID for the edit."""
        return str(uuid.uuid4())

    def _serialize_value(self, value: Any) -> Optional[str]:
        """Convert a value to string for storage."""
        if value is None:
            return None
        if isinstance(value, (datetime,)):
            return value.isoformat()
        if isinstance(value, (dict, list)):
            import json
            return json.dumps(value, default=str)
        return str(value)

    def _insert_edit(self, table_name: str, record_id: int, field_name: Optional[str],
                     action: str, old_value: Any, new_value: Any) -> int:
        """Insert an edit record and return its ID."""
        edit_uuid = self._generate_uuid()

        self.cursor.execute("""
            INSERT INTO edits (
                edit_uuid, table_name, record_id, field_name, action,
                old_value, new_value, source, user_id, editor_ip,
                editor_user_agent, editor_info, applied_at
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, NOW())
        """, (
            edit_uuid,
            table_name,
            record_id,
            field_name,
            action,
            self._serialize_value(old_value),
            self._serialize_value(new_value),
            self.source,
            self.user_id,
            self.editor_ip,
            self.editor_user_agent,
            self.editor_info
        ))

        return self.cursor.lastrowid

    def log_insert(self, table_name: str, record_id: int, record_data: Dict[str, Any]) -> int:
        """
        Log an INSERT operation.

        Args:
            table_name: Name of the table
            record_id: ID of the inserted record
            record_data: Dict of field names to values

        Returns:
            Edit ID
        """
        if table_name not in self.TRACKED_TABLES:
            return None

        return self._insert_edit(
            table_name=table_name,
            record_id=record_id,
            field_name=None,
            action='INSERT',
            old_value=None,
            new_value=record_data
        )

    def log_update(self, table_name: str, record_id: int, field_name: str,
                   old_value: Any, new_value: Any) -> int:
        """
        Log an UPDATE operation for a single field.

        Args:
            table_name: Name of the table
            record_id: ID of the updated record
            field_name: Name of the field that changed
            old_value: Previous value
            new_value: New value

        Returns:
            Edit ID
        """
        if table_name not in self.TRACKED_TABLES:
            return None

        # Don't log if value didn't actually change
        if self._serialize_value(old_value) == self._serialize_value(new_value):
            return None

        return self._insert_edit(
            table_name=table_name,
            record_id=record_id,
            field_name=field_name,
            action='UPDATE',
            old_value=old_value,
            new_value=new_value
        )

    def log_updates(self, table_name: str, record_id: int,
                    old_record: Dict[str, Any], new_record: Dict[str, Any]) -> List[int]:
        """
        Log UPDATE operations for multiple fields by comparing old and new records.

        Args:
            table_name: Name of the table
            record_id: ID of the updated record
            old_record: Dict of old field values
            new_record: Dict of new field values

        Returns:
            List of edit IDs
        """
        edit_ids = []

        for field_name, new_value in new_record.items():
            old_value = old_record.get(field_name)
            edit_id = self.log_update(table_name, record_id, field_name, old_value, new_value)
            if edit_id:
                edit_ids.append(edit_id)

        return edit_ids

    def log_delete(self, table_name: str, record_id: int, record_data: Dict[str, Any]) -> int:
        """
        Log a DELETE operation.

        Args:
            table_name: Name of the table
            record_id: ID of the deleted record
            record_data: Dict of field names to values (snapshot before deletion)

        Returns:
            Edit ID
        """
        if table_name not in self.TRACKED_TABLES:
            return None

        return self._insert_edit(
            table_name=table_name,
            record_id=record_id,
            field_name=None,
            action='DELETE',
            old_value=record_data,
            new_value=None
        )

    def get_edits_since(self, since_id: int = 0, source: str = None,
                        limit: int = 1000) -> List[Dict]:
        """
        Get edits since a given edit ID.

        Args:
            since_id: Return edits with ID > since_id
            source: Filter by source ('local', 'website', 'crawl')
            limit: Maximum number of edits to return

        Returns:
            List of edit dicts
        """
        sql = """
            SELECT id, edit_uuid, table_name, record_id, field_name, action,
                   old_value, new_value, source, user_id, editor_ip,
                   editor_user_agent, editor_info, created_at, applied_at
            FROM edits
            WHERE id > %s
        """
        params = [since_id]

        if source:
            sql += " AND source = %s"
            params.append(source)

        sql += " ORDER BY id ASC LIMIT %s"
        params.append(limit)

        self.cursor.execute(sql, params)

        columns = [
            'id', 'edit_uuid', 'table_name', 'record_id', 'field_name', 'action',
            'old_value', 'new_value', 'source', 'user_id', 'editor_ip',
            'editor_user_agent', 'editor_info', 'created_at', 'applied_at'
        ]

        edits = []
        for row in self.cursor.fetchall():
            edit = dict(zip(columns, row))
            # Convert datetime objects to ISO strings for serialization
            if edit['created_at']:
                edit['created_at'] = edit['created_at'].isoformat()
            if edit['applied_at']:
                edit['applied_at'] = edit['applied_at'].isoformat()
            edits.append(edit)

        return edits

    def get_record_history(self, table_name: str, record_id: int) -> List[Dict]:
        """
        Get edit history for a specific record.

        Args:
            table_name: Name of the table
            record_id: ID of the record

        Returns:
            List of edit dicts, newest first
        """
        self.cursor.execute("""
            SELECT e.id, e.edit_uuid, e.field_name, e.action,
                   e.old_value, e.new_value, e.source, e.editor_info,
                   e.created_at, u.display_name as user_name, u.email as user_email
            FROM edits e
            LEFT JOIN users u ON e.user_id = u.id
            WHERE e.table_name = %s AND e.record_id = %s
            ORDER BY e.created_at DESC
        """, (table_name, record_id))

        columns = [
            'id', 'edit_uuid', 'field_name', 'action', 'old_value', 'new_value',
            'source', 'editor_info', 'created_at', 'user_name', 'user_email'
        ]

        edits = []
        for row in self.cursor.fetchall():
            edit = dict(zip(columns, row))
            if edit['created_at']:
                edit['created_at'] = edit['created_at'].isoformat()
            edits.append(edit)

        return edits

    def apply_edit(self, edit: Dict) -> bool:
        """
        Apply an edit from another source to the local database.

        Args:
            edit: Edit dict from get_edits_since()

        Returns:
            True if successfully applied
        """
        table_name = edit['table_name']
        record_id = edit['record_id']
        action = edit['action']

        if table_name not in self.TRACKED_TABLES:
            return False

        try:
            if action == 'UPDATE':
                field_name = edit['field_name']
                new_value = edit['new_value']

                # Apply the update
                self.cursor.execute(
                    f"UPDATE {table_name} SET {field_name} = %s WHERE id = %s",
                    (new_value, record_id)
                )

            elif action == 'DELETE':
                self.cursor.execute(
                    f"DELETE FROM {table_name} WHERE id = %s",
                    (record_id,)
                )

            elif action == 'INSERT':
                # INSERT is more complex - need to handle the full record
                # For now, skip INSERT syncs (they should be handled differently)
                return False

            # Record that we applied this edit
            self.cursor.execute("""
                INSERT INTO edits (
                    edit_uuid, table_name, record_id, field_name, action,
                    old_value, new_value, source, editor_info, applied_at
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, NOW())
                ON DUPLICATE KEY UPDATE applied_at = NOW()
            """, (
                edit['edit_uuid'],
                table_name,
                record_id,
                edit.get('field_name'),
                action,
                edit.get('old_value'),
                edit.get('new_value'),
                edit['source'],
                f"synced from {edit['source']}"
            ))

            return True

        except Exception as e:
            print(f"Error applying edit {edit['edit_uuid']}: {e}")
            return False


def get_edit_logger(cursor, connection, source: str = 'local',
                    editor_info: str = None) -> EditLogger:
    """
    Factory function to create an EditLogger.

    Args:
        cursor: Database cursor
        connection: Database connection
        source: Origin of edits ('local', 'website', or 'crawl')
        editor_info: Additional context string

    Returns:
        EditLogger instance
    """
    return EditLogger(cursor, connection, source=source, editor_info=editor_info)

File: src/api/edit_logger.php
<?php
/**
 * Edit Logger
 *
 * Logs all changes to core tables for sync and audit purposes.
 * Each edit is stored as an immutable log entry with a UUID for global uniqueness.
 */

require_once __DIR__ . '/config.php';

class EditLogger {
    private $pdo;
    private $source;
    private $editorInfo;
    private $userId;
    private $editorIp;
    private $editorUserAgent;

    // Tables that should have edits logged
    private const TRACKED_TABLES = [
        'locations', 'location_alternate_names', 'location_tags',
        'websites', 'website_urls', 'website_locations', 'website_tags',
        'events', 'event_occurrences', 'event_urls', 'event_tags',
        'tags', 'tag_rules'
    ];

    /**
     * Initialize the edit logger.
     *
     * @param PDO $pdo Database connection
     * @param string $source Origin of edits ('local', 'website', or 'crawl')
     * @param string|null $editorInfo Additional context (e.g., 'admin', 'api')
     */
    public function __construct(PDO $pdo, string $source = 'website', ?string $editorInfo = null) {
        $this->pdo = $pdo;
        $this->source = $source;
        $this->editorInfo = $editorInfo;
        $this->userId = null;
        $this->editorIp = null;
        $this->editorUserAgent = null;
    }

    /**
     * Set user context for attribution.
     */
    public function setUserContext(?int $userId = null, ?string $ip = null, ?string $userAgent = null): void {
        $this->userId = $userId;
        $this->editorIp = $ip;
        $this->editorUserAgent = $userAgent;
    }

    /**
     * Auto-detect user context from request.
     */
    public function setUserContextFromRequest(): void {
        $this->editorIp = $_SERVER['REMOTE_ADDR'] ?? null;
        $this->editorUserAgent = $_SERVER['HTTP_USER_AGENT'] ?? null;

        // Check for logged-in user in session
        if (session_status() === PHP_SESSION_ACTIVE && isset($_SESSION['user_id'])) {
            $this->userId = $_SESSION['user_id'];
        }
    }

    /**
     * Generate a UUID v4.
     */
    private function generateUuid(): string {
        $data = random_bytes(16);
        $data[6] = chr(ord($data[6]) & 0x0f | 0x40);
        $data[8] = chr(ord($data[8]) & 0x3f | 0x80);
        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }

    /**
     * Serialize a value for storage.
     */
    private function serializeValue($value): ?string {
        if ($value === null) {
            return null;
        }
        if (is_array($value) || is_object($value)) {
            return json_encode($value);
        }
        if ($value instanceof DateTime) {
            return $value->format('c');
        }
        return (string)$value;
    }

    /**
     * Check if a table is tracked.
     */
    private function isTrackedTable(string $tableName): bool {
        return in_array($tableName, self::TRACKED_TABLES);
    }

    /**
     * Insert an edit record.
     */
    private function insertEdit(
        string $tableName,
        int $recordId,
        ?string $fieldName,
        string $action,
        $oldValue,
        $newValue
    ): ?int {
        $editUuid = $this->generateUuid();

        $stmt = $this->pdo->prepare("
            INSERT INTO edits (
                edit_uuid, table_name, record_id, field_name, action,
                old_value, new_value, source, user_id, editor_ip,
                editor_user_agent, editor_info, applied_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, NOW())
        ");

        $stmt->execute([
            $editUuid,
            $tableName,
            $recordId,
            $fieldName,
            $action,
            $this->serializeValue($oldValue),
            $this->serializeValue($newValue),
            $this->source,
            $this->userId,
            $this->editorIp,
            $this->editorUserAgent ? substr($this->editorUserAgent, 0, 500) : null,
            $this->editorInfo
        ]);

        return (int)$this->pdo->lastInsertId();
    }

    /**
     * Log an INSERT operation.
     *
     * @param string $tableName Name of the table
     * @param int $recordId ID of the inserted record
     * @param array $recordData Associative array of field values
     * @return int|null Edit ID
     */
    public function logInsert(string $tableName, int $recordId, array $recordData): ?int {
        if (!$this->isTrackedTable($tableName)) {
            return null;
        }

        return $this->insertEdit(
            $tableName,
            $recordId,
            null,
            'INSERT',
            null,
            $recordData
        );
    }

    /**
     * Log an UPDATE operation for a single field.
     *
     * @param string $tableName Name of the table
     * @param int $recordId ID of the updated record
     * @param string $fieldName Name of the field that changed
     * @param mixed $oldValue Previous value
     * @param mixed $newValue New value
     * @return int|null Edit ID
     */
    public function logUpdate(
        string $tableName,
        int $recordId,
        string $fieldName,
        $oldValue,
        $newValue
    ): ?int {
        if (!$this->isTrackedTable($tableName)) {
            return null;
        }

        // Don't log if value didn't actually change
        if ($this->serializeValue($oldValue) === $this->serializeValue($newValue)) {
            return null;
        }

        return $this->insertEdit(
            $tableName,
            $recordId,
            $fieldName,
            'UPDATE',
            $oldValue,
            $newValue
        );
    }

    /**
     * Log UPDATE operations for multiple fields by comparing old and new records.
     *
     * @param string $tableName Name of the table
     * @param int $recordId ID of the updated record
     * @param array $oldRecord Old field values
     * @param array $newRecord New field values
     * @return array List of edit IDs
     */
    public function logUpdates(
        string $tableName,
        int $recordId,
        array $oldRecord,
        array $newRecord
    ): array {
        $editIds = [];

        foreach ($newRecord as $fieldName => $newValue) {
            $oldValue = $oldRecord[$fieldName] ?? null;
            $editId = $this->logUpdate($tableName, $recordId, $fieldName, $oldValue, $newValue);
            if ($editId !== null) {
                $editIds[] = $editId;
            }
        }

        return $editIds;
    }

    /**
     * Log a DELETE operation.
     *
     * @param string $tableName Name of the table
     * @param int $recordId ID of the deleted record
     * @param array $recordData Snapshot of record before deletion
     * @return int|null Edit ID
     */
    public function logDelete(string $tableName, int $recordId, array $recordData): ?int {
        if (!$this->isTrackedTable($tableName)) {
            return null;
        }

        return $this->insertEdit(
            $tableName,
            $recordId,
            null,
            'DELETE',
            $recordData,
            null
        );
    }

    /**
     * Get edits since a given edit ID.
     *
     * @param int $sinceId Return edits with ID > sinceId
     * @param string|null $source Filter by source
     * @param int $limit Maximum number of edits
     * @return array List of edit records
     */
    public function getEditsSince(int $sinceId = 0, ?string $source = null, int $limit = 1000): array {
        $sql = "
            SELECT id, edit_uuid, table_name, record_id, field_name, action,
                   old_value, new_value, source, user_id, editor_ip,
                   editor_user_agent, editor_info, created_at, applied_at
            FROM edits
            WHERE id > ?
        ";
        $params = [$sinceId];

        if ($source !== null) {
            $sql .= " AND source = ?";
            $params[] = $source;
        }

        // LIMIT value is already sanitized as int, so interpolate directly
        // (PDO emulation mode can't bind integers for LIMIT)
        $sql .= " ORDER BY id ASC LIMIT " . $limit;

        $stmt = $this->pdo->prepare($sql);
        $stmt->execute($params);

        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }

    /**
     * Get edit history for a specific record.
     *
     * @param string $tableName Name of the table
     * @param int $recordId ID of the record
     * @return array List of edit records, newest first
     */
    public function getRecordHistory(string $tableName, int $recordId): array {
        $stmt = $this->pdo->prepare("
            SELECT e.id, e.edit_uuid, e.field_name, e.action,
                   e.old_value, e.new_value, e.source, e.editor_info,
                   e.created_at, u.display_name as user_name, u.email as user_email
            FROM edits e
            LEFT JOIN users u ON e.user_id = u.id
            WHERE e.table_name = ? AND e.record_id = ?
            ORDER BY e.created_at DESC
        ");

        $stmt->execute([$tableName, $recordId]);
        return $stmt->fetchAll(PDO::FETCH_ASSOC);
    }

    /**
     * Get the last synced edit ID for a source.
     *
     * @param string $source 'local' or 'website'
     * @return int|null Last synced edit ID
     */
    public function getLastSyncedEditId(string $source): ?int {
        $stmt = $this->pdo->prepare("
            SELECT last_synced_edit_id FROM sync_state WHERE source = ?
        ");
        $stmt->execute([$source]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        return $row ? (int)$row['last_synced_edit_id'] : null;
    }

    /**
     * Update the last synced edit ID for a source.
     *
     * @param string $source 'local' or 'website'
     * @param int $editId Last synced edit ID
     */
    public function updateLastSyncedEditId(string $source, int $editId): void {
        $stmt = $this->pdo->prepare("
            INSERT INTO sync_state (source, last_synced_edit_id, last_sync_at)
            VALUES (?, ?, NOW())
            ON DUPLICATE KEY UPDATE last_synced_edit_id = ?, last_sync_at = NOW()
        ");
        $stmt->execute([$source, $editId, $editId]);
    }
}

/**
 * Create an EditLogger instance with database connection.
 *
 * @param string $source Origin of edits ('local', 'website', or 'crawl')
 * @param string|null $editorInfo Additional context
 * @return EditLogger
 */
function getEditLogger(string $source = 'website', ?string $editorInfo = null): EditLogger {
    $pdo = new PDO(
        "mysql:host=" . DB_HOST . ";dbname=" . DB_NAME . ";charset=utf8mb4",
        DB_USER,
        DB_PASS,
        [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]
    );

    $logger = new EditLogger($pdo, $source, $editorInfo);
    $logger->setUserContextFromRequest();

    return $logger;
}
